polynote-kernel/src/main/scala/polynote/kernel/lang/scal/ScalaInterpreter.scala

```diff
@@ -195,7 +195,7 @@ class ScalaInterpreter(
               ioValues =>
                 Stream(
                   maybeResultQ.dequeue.unNoneTerminate,
-                  Stream.eval(ioValues).flatMap(Stream.emits).onFinalize(maybeResultQ.enqueue1(None))
+                  Stream.eval(ioValues).flatMap(Stream.emits) ++ Stream.eval(maybeResultQ.enqueue1(None)).drain
```

Comment:
This is because there were intermittent test failures where not all the output seemed to make it through before the queue was closed. I think there's a path toward cleaning up this anti-pattern and I'll put that on my TODO list... but for now I think this will help

Reply:
Hm, do you think we should look for other `Stream.onFinalize` usages (if any) and fix them as well?

I wonder whether for ease of refactoring we should consider a utility method like 
```
// TODO: change this finalizer
def todoChangeThisStreamFinalizer(s: Stream[IO, Result], finalizer: IO[Unit]) = {
    s ++ Stream.eval(finalizer).drain
}
```

This way once we figure out how to clean things up we'll know where to look. The deliberately ugly name is to encourage cleaning it sooner rather than later 😛

Reply:
I think the semantics of `onFinalize` are just... different. If you look at what it does it introduces a `bracket` construct, so the finalizer just isn't part of the stream (i.e. someone downstream waiting for this stream to terminate won't necessarily wait for the finalizer, so there could be a race condition).

I think the solution isn't to try and switch to finalize, but rather to try and get rid of all of these `Queue`s and `EnqueueSome`s and `unNoneTerminate`s that we have everywhere. It's really a silly way of doing it... we just want a special value (pretty much always a `Result`) that means `The End`, so we can terminate a stream when we see it. Right now we make that special value be `None`, because there's an API for that, but it also means wrapping everything else into `Option` for the sole purpose of having `None` be the EOF marker.

But we already have a start marker `Result` – `ClearResults` (which could just as easily be called `StartResults`. So why not just our own EOF marker? The API support for that isn't baked into fs2 as far as I can tell (or at least it's not clear which methods have "halt" semantics and which have "interrupt" semantics). But I think implementing that mechanism could simplify a lot of our code pretty dramatically... I always cringe when I see `EnqueueSome` and friends.

Reply:
P.S. I think an EOF marker `Result` would also be useful on the client side!